pandas/core/frame.py

```diff
@@ -647,7 +647,7 @@ def __init__(
         elif isinstance(data, (np.ndarray, Series, Index)):
             if data.dtype.names:
                 # i.e. numpy structured array
-                data = cast(np.ndarray, data)
+                data = cast(np.recarray, data)
```

Comment:
how is this guaranteed?

```
>>> dt = np.dtype([('name', np.str_, 16), ('grades', np.float64, (2,))])
>>> 
>>> arr = np.array([], dtype=dt)
>>> 
>>> arr.dtype.names
('name', 'grades')
>>> 
>>> type(arr)
<class 'numpy.ndarray'>
```

Reply:
> how is this guaranteed?

it appears not, constructing a DataFrame from a compound dtype works and goes through here.

```
> /home/simon/pandas/pandas/core/frame.py(651)__init__()
    649                 # i.e. numpy structured array
    650                 breakpoint()
--> 651                 data = cast(np.ndarray, data)
    652                 mgr = rec_array_to_mgr(
    653                     data,

ipdb> data
array([(1., 2), (3., 4)], dtype=[('x', '<f8'), ('y', '<i8')])
ipdb> type(data)
<class 'numpy.ndarray'>
```

Reply:
huh, i learned something today.